src/pemfile.rs

```diff
@@ -36,6 +43,51 @@ pub enum Item {
     Crl(CertificateRevocationListDer<'static>),
 }
 
+/// Errors that may arise when parsing the contents of a PEM file
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// a section is missing its "END marker" line
+    MissingSectionEnd {
+        /// the expected "END marker" line that was not found
+        end_marker: Vec<u8>,
+    },
+
+    /// syntax error found in the line that starts a new section
+    IllegalSectionStart {
+        /// line that contains the syntax error
+        line: Vec<u8>,
+    },
+
+    /// base64 decode error
+    Base64Decode(String),
+}
+
+/// Extract and decode the next PEM section from `input`
+///
+/// - `Ok(None)` is returned if there is no PEM section to read from `input`
+/// - Syntax errors and decoding errors produce a `Err(...)`
+/// - Otherwise each decoded section is returned with a `Ok(Some((Item::..., remainder)))` where
+///   `remainder` is the part of the `input` that follows the returned section
+pub fn read_one_from_slice(mut input: &[u8]) -> Result<Option<(Item, &[u8])>, Error> {
+    let mut b64buf = Vec::with_capacity(1024);
+    let mut section = None::<(Vec<_>, Vec<_>)>;
+
+    loop {
+        let next_line = if let Some(index) = input.iter().position(|byte| *byte == b'\n') {
```

Comment:
Out of curiosity, were these changes benchmarked? I was also making some changes to the implementation here and observed that without std's usage of `memchr` or the `memchr` crate, performance tanks by about ~2x in my branch due to the performance looking for line endings.

Reply:
Is that with the `std` feature disabled? It seems odd that they are so close but its not a big deal.